pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go

```diff
@@ -67,48 +70,82 @@ func (ba *BalancedAllocation) ScoreExtensions() framework.ScoreExtensions {
 }
 
 // NewBalancedAllocation initializes a new plugin and returns it.
-func NewBalancedAllocation(_ runtime.Object, h framework.Handle) (framework.Plugin, error) {
+func NewBalancedAllocation(baArgs runtime.Object, h framework.Handle) (framework.Plugin, error) {
+	args, ok := baArgs.(*config.NodeResourcesBalancedAllocationArgs)
+	if !ok {
+		return nil, fmt.Errorf("want args to be of type NodeResourcesBalancedAllocationArgs, got %T", baArgs)
+	}
+
+	if err := validation.ValidateNodeResourcesBalancedAllocationArgs(nil, args); err != nil {
+		return nil, err
+	}
+
+	resToWeightMap := make(resourceToWeightMap)
+	for _, resource := range (*args).Resources {
+		resToWeightMap[v1.ResourceName(resource.Name)] = resource.Weight
+	}
+
 	return &BalancedAllocation{
 		handle: h,
 		resourceAllocationScorer: resourceAllocationScorer{
 			BalancedAllocationName,
 			balancedResourceScorer,
-			defaultRequestedRatioResources,
+			resToWeightMap,
 		},
 	}, nil
 }
 
-// todo: use resource weights in the scorer function
 func balancedResourceScorer(requested, allocable resourceToValueMap, includeVolumes bool, requestedVolumes int, allocatableVolumes int) int64 {
-	cpuFraction := fractionOfCapacity(requested[v1.ResourceCPU], allocable[v1.ResourceCPU])
-	memoryFraction := fractionOfCapacity(requested[v1.ResourceMemory], allocable[v1.ResourceMemory])
-	// This to find a node which has most balanced CPU, memory and volume usage.
-	if cpuFraction >= 1 || memoryFraction >= 1 {
-		// if requested >= capacity, the corresponding host should never be preferred.
-		return 0
+	resourceToFraction := make(map[v1.ResourceName]float64)
+	var totalFraction float64
+	for name, value := range requested {
+		if _, ok := allocable[name]; ok {
+			fraction := fractionOfCapacity(value, allocable[name])
+			if fraction >= 1 {
+				// if requested >= capacity, the corresponding host should never be preferred.
+				return 0
+			}
+			totalFraction = totalFraction + fraction
+			resourceToFraction[name] = fraction
+		} else {
+			klog.InfoS("requested resource not found for node score calculation", "resource", name)
+			return 0
```

Comment:
I don't think it's appropriate to return 0 if a Pod doesn't request ALL resources configured in the args.

Reply:
maybe I read it wrong, I was thinking that when a pod request something in the config while this node cannot provide, then this node is not eligible to schedule.

Reply:
Aldo is right, the `requested` and `allocable` are instantiated with the same size of passed resToWeight map:

```go
for resource := range r.resourceToWeightMap {
	allocatable[resource], requested[resource] = calculateResourceAllocatableRequest(nodeInfo, pod, resource)
}
```

So let's suppose the resToWeight map is specified as {cpu, mem, gpu}:

### Case 1: a Pod doesn't request gpu, and the evaluating Node doesn't have gpu resource

- `ok` here is always true
- when it comes to cpu/mem, it's calculated properly - accumulated in resourceToFraction
- when it comes to gpu, it returns 0 immediately, **which doesn't look right**, as it eliminates the differentiator of other resources (cpu/mem).

### Case 2: a Pod doesn't request gpu, and the evaluating Node has gpu resource

- `ok` here is always true
- when it comes to cpu/mem, it calculated properly - accumulated in resourceToFraction
- when it comes to gpu, resourceToFraction[gpu] = 0, which is as expected

### Case 3: a Pod requests gpu, but the evaluating Node doesn't have gpu resource

As @alculquicondor mentioned, we shouldn't come here as it would have already failed in Filter.

With both case 1 and case 2 mentioned, if a Pod doesn't request particular extended resources (e.g., gpu), the question is: how do we fairly score the with-gpu and without-gpu nodes? In current implementation, it returns 0 for all nodes.

My 2 cents is: **for extended resources**, if the request is 0, don't instantiate its entry in the `requested` [map](https://github.com/kubernetes/kubernetes/blob/3b73c187c74a1d758902b6ff279735c36d87d0e9/pkg/scheduler/framework/plugins/noderesources/resource_allocation.go#L55), then when it's scored afterwards, the non-requested extended resource won't be iterated at all. WDYT?

Reply:
The above statement is also valid for other score plugins: least_allocated and most_allocated since they both use PluginArgs to define extended resources.

The symptom for these 2 plugins is: if an extended resource is given in `{Least|Mode}AllocatedArgs`, and a Pod doesn't request that extended resource, it won't be scored a MaxNodeScore (100) in any case. For example, `{Least|Mode}AllocatedArgs` is given as {cpu, mem, gpu}, a Pod without requesting gpu can get a maximum node score as 67: (100+100) / 3.

Like the comment above, we may want to ensure the MaxNodeScore is always 100 - no matter a Pod request a defined extended resource or not.

Reply:
Actually, for case 2, a score lower than 100 is desired. Basically, we don't want to "waste" that extended resource.

For case 1, we should ensure that the score can get to 100.

Reply:
> Actually, for case 2, a score lower than 100 is desired. Basically, we don't want to "waste" that extended resource.

The favorness result won't change for either formula. It's just whether the score should be diluted when an extended res is defined but the Pod doesn't request it. I will leave it to open discussion, and it shouldn't block this PR.

Reply:
It should be diluted. If a node has GPUs, but the pod doesn't request it, it should score lower than a node that doesn't have GPUs.
However, remember that the plugins don't actually look at the requests of the pod, but rather the sum of requests of already running pods and the pod we are scheduling.

Reply:
> It should be diluted.

It worths a discussion. Let me give an example: suppose we have 3 nodes, and {cpu, mem, gpu} are registered through pluginArgs and all weighted 1. And below is the current requested/capacity usage:

|       | cpu | mem | gpu |
|-------|-----|-----|-----|
| node1 | 4/8 | 4/8 | 2/8 |
| node2 | 4/8 | 4/8 | 4/8 |
| node3 | 4/8 | 4/8 | 0/0 (no gpus) |

Now here comes a best-efforts Pod, it will be scored like this:

|       | cpu | mem | gpu                                         |
|-------|-----|-----|---------------------------------------------|
| node1 | 50  | 50  | 75 (least requested)<br>25 (most requested) |
| node2 | 50  | 50  | 50                                          |
| node3 | 50  | 50  | 0                                           |

node3 is not selected for least/most requested plugin, but shouldn't it be favored in both cases? As it would leave the cpu/mem room so another Pod requests gpu can get more chances to be scheduled to node1/node2.

Reply:
ok, for the case where the node has no GPUs, it shoudn't be diluted. Simply GPU shouldn't be accounted.

My point for dilution was for the other case, where the node has GPUs, but the pod doesn't request it.

Reply:
> My point for dilution was for the other case, where the node has GPUs, but the pod doesn't request it.

Agree, which is case node1/node2 in the above example.

Let me pull this concern out to a new issue for better brainstorming. (as it applies to all plugins that allow customizing extended resource)

Reply:
so....

/hold